File: src/main/java/com/morph/engine/math/Vector4f.java
package com.morph.engine.math;

public class Vector4f {
	private float x, y, z, w;
	public Vector4f(float x, float y, float z, float w) {
		this.x = x;
		this.y = y;
		this.z = z;
		this.w = w;
	}
	
	/**
	 * Clones a Vector4f.
	 */
	public Vector4f(Vector4f v) {
		this(v.getX(), v.getY(), v.getZ(), v.getW());
	}
	
	/**
	 * Creates a new Vector4f from a Vector3f and a supplied z value.
	 * @param v Vector3f containing new x and y values.
	 */
	public Vector4f(Vector3f v, float w) {
		this(v.getX(), v.getY(), v.getZ(), w);
	}
	
	/**
	 * Creates a new Vector4f from a Vector3f by using the x and y values of the Vector3f.
	 * @param v Vector3f containing new x and y values.
	 */
	public Vector4f(Vector3f v) {
		this(v, 0);
	}
	
	public Vector4f() {
		this.x = 0;
		this.y = 0;
		this.z = 0;
		this.w = 0;
	}
	
	public Vector4f(Vector2f xy, int z, int w) {
		this.x = xy.getX();
		this.y = xy.getY();
		this.z = z;
		this.w = w;
	}

	public String toString() {
		return "Vector4f(" + x + ", " + y + ", " + z + ", " + w + ")";
	}
	
	public Vector4f add(Vector4f v) {
		return new Vector4f(x + v.getX(), y + v.getY(), z + v.getZ(), w + v.getW());
	}
	
	public Vector4f sub(Vector4f v) {
		return new Vector4f(x - v.getX(), y - v.getY(), z - v.getZ(), w - v.getW());
	}
	
	public Vector4f scale(float k) {
		return new Vector4f(x * k, y * k, z * k, w * k);
	}
	
	public Vector4f invScale(float k) {
		return new Vector4f(x / k, y / k, z / k, w / k);
	}

	public float getX() {
		return x;
	}
	
	public float getY() {
		return y;
	}
	
	public float getZ() {
		return z;
	}
	
	public float getW() {
		return w;
	}
	
	public void setX(float x) {
		this.x = x;
	}
	
	public void setY(float y) {
		this.y = y;
	}
	
	public void setZ(float z) {
		this.z = z;
	}
	
	public void setW(float w) {
		this.w = w;
	}
	
	public void set(float x, float y, float z, float w) {
		this.x = x;
		this.y = y;
		this.z = z;
		this.w = w;
	}
	
	public void set(Vector4f v) {
		this.x = v.getX();
		this.y = v.getY();
		this.z = v.getZ();
		this.w = v.getW();
	}

	public Vector3f getXYZ() {
		return new Vector3f(x, y, z);
	}

	public Vector2f getXY() {
		return new Vector2f(x, y);
	}
}

File: src/main/kotlin/com/morph/engine/math/Vector4f.kt
package com.morph.engine.math

import java.util.function.Function

class Vector4f(var x : Float = 0f, var y : Float = 0f, var z : Float = 0f, var w : Float = 0f) {
    val xyz: Vector3f
        get() = Vector3f(x, y, z)

    val xy: Vector2f
        get() = Vector2f(x, y)

    /**
     * Clones a Vector4f.
     */
    constructor(v: Vector4f) : this(v.x, v.y, v.z, v.w) {}

    /**
     * Creates a new Vector4f from a Vector3f and a supplied z value.
     * @param v Vector3f containing new x and y values.
     */
    @JvmOverloads constructor(v: Vector3f, w: Float = 0f) : this(v.x, v.y, v.z, w)

    constructor(xy: Vector2f, z: Float, w: Float) : this(xy.x, xy.y, z, w)

    override fun toString(): String {
        return "Vector4f($x, $y, $z, $w)"
    }

    @JvmName("add")
    operator fun plus(v: Vector4f): Vector4f {
        return Vector4f(x + v.x, y + v.y, z + v.z, w + v.w)
    }

    @JvmName("sub")
    operator fun minus(v: Vector4f): Vector4f {
        return Vector4f(x - v.x, y - v.y, z - v.z, w - v.w)
    }

    @JvmName("scale")
    operator fun times(k: Float): Vector4f {
        return Vector4f(x * k, y * k, z * k, w * k)
    }

    @JvmName("invScale")
    operator fun div(k: Float): Vector4f {
        return Vector4f(x / k, y / k, z / k, w / k)
    }

    operator fun set(x: Float, y: Float, z: Float, w: Float) {
        this.x = x
        this.y = y
        this.z = z
        this.w = w
    }

    fun set(v: Vector4f) {
        this.x = v.x
        this.y = v.y
        this.z = v.z
        this.w = v.w
    }

    fun map(`fun`: Function<Float, Float>): Vector4f {
        return Vector4f(`fun`.apply(x), `fun`.apply(y), `fun`.apply(z), `fun`.apply(w))
    }
}
/**
 * Creates a new Vector4f from a Vector3f by using the x and y values of the Vector3f.
 * @param v Vector3f containing new x and y values.
 */
